benchmarks/foveation/main.cpp

```diff
@@ -0,0 +1,771 @@
+{
+    auto queue = GetDevice()->GetGraphicsQueue();
+
+    const grfx::FormatDesc* formatDesc = grfx::GetFormatDescription(image->GetFormat());
+    const uint32_t          width      = image->GetWidth();
+    const uint32_t          height     = image->GetHeight();
+
+    // Create a buffer that will hold the swapchain image's texels.
+    // Increase its size by a factor of 2 to ensure that a larger-than-needed
+    // row pitch will not overflow the buffer.
+    uint64_t bufferSize = 2ull * formatDesc->bytesPerTexel * width * height;
+
+    grfx::BufferPtr        screenshotBuf = nullptr;
+    grfx::BufferCreateInfo bufferCi      = {};
+    bufferCi.size                        = bufferSize;
+    bufferCi.initialState                = grfx::RESOURCE_STATE_COPY_DST;
+    bufferCi.usageFlags.bits.transferDst = 1;
+    bufferCi.memoryUsage                 = grfx::MEMORY_USAGE_GPU_TO_CPU;
+    PPX_CHECKED_CALL(GetDevice()->CreateBuffer(&bufferCi, &screenshotBuf));
+
+    // We wait for idle so that we can avoid tracking swapchain fences.
+    // It's not ideal, but we won't be taking screenshots in
+    // performance-critical scenarios.
+    PPX_CHECKED_CALL(queue->WaitIdle());
+
+    // Copy the swapchain image into the buffer.
+    grfx::CommandBufferPtr cmdBuf;
+    PPX_CHECKED_CALL(queue->CreateCommandBuffer(&cmdBuf, 0, 0));
+
+    grfx::ImageToBufferOutputPitch outPitch;
+    cmdBuf->Begin();
+    {
+        cmdBuf->TransitionImageLayout(image, PPX_ALL_SUBRESOURCES, resourceState, grfx::RESOURCE_STATE_COPY_SRC);
+
+        grfx::ImageToBufferCopyInfo bufCopyInfo = {};
+        bufCopyInfo.extent                      = {width, height, 0};
+        outPitch                                = cmdBuf->CopyImageToBuffer(&bufCopyInfo, image, screenshotBuf);
+
+        cmdBuf->TransitionImageLayout(image, PPX_ALL_SUBRESOURCES, grfx::RESOURCE_STATE_COPY_SRC, resourceState);
+    }
+    cmdBuf->End();
+
+    grfx::SubmitInfo submitInfo   = {};
+    submitInfo.commandBufferCount = 1;
+    submitInfo.ppCommandBuffers   = &cmdBuf;
+    queue->Submit(&submitInfo);
+
+    // Wait for the copy to be finished.
+    queue->WaitIdle();
+
+    // Export to PPM.
+    unsigned char* texels = nullptr;
+    screenshotBuf->MapMemory(0, (void**)&texels);
+
+    PPX_CHECKED_CALL(ExportToPPM(filepath, image->GetFormat(), texels, width, height, outPitch.rowPitch));
+
+    screenshotBuf->UnmapMemory();
+
+    // Clean up temporary resources.
+    GetDevice()->DestroyBuffer(screenshotBuf);
+    queue->DestroyCommandBuffer(cmdBuf);
+}
+
+void FoveationBenchmarkApp::Render()
+{
+    uint32_t imageIndex = UINT32_MAX;
+    PPX_CHECKED_CALL(GetSwapchain()->AcquireNextImage(UINT64_MAX, mSync.imageAcquiredSemaphore, mSync.imageAcquiredFence, &imageIndex));
+
+    PPX_LOG_INFO("FoveationBenchmarkApp::Render imageIndex:" << imageIndex);
+
+    // Wait for and reset render complete fence
+    PPX_CHECKED_CALL(mSync.postCompleteFence->WaitAndReset());
```

Comment:
I think we need to move this to before AcquireNextImage. See https://github.com/google/bigwheels/pull/445